java-checks/src/main/java/org/sonar/java/checks/serialization/SerializableObjectInSessionCheck.java

```diff
@@ -44,8 +47,45 @@
   protected void onMethodInvocationFound(MethodInvocationTree mit) {
     ExpressionTree argument = mit.arguments().get(1);
     Type type = argument.symbolType();
-    if (!type.isPrimitive() && !type.isSubtypeOf("java.io.Serializable")) {
-      reportIssue(argument, "Make \"" + type + "\" serializable or don't store it in the session.");
+    if (!isSerializable(type)) {
+      String andParameters = isParametrized(type) ? " and its parameters" : "";
+      reportIssue(argument, "Make \"" + type + "\"" + andParameters + " serializable or don't store it in the session.");
     }
   }
+
+  private boolean isSerializable(Type type) {
+    if (type.isPrimitive()) {
+      return true;
+    }
+    if (isSerializableArray(type)) {
+      return true;
+    }
+    if (isParametrized(type)) {
+      return isSerializableParametrized((ParametrizedTypeJavaType) type);
+    }
+    return type.isSubtypeOf("java.io.Serializable");
+  }
+
+  private boolean isSerializableArray(Type type) {
+    if (type instanceof ArrayJavaType) {
+      ArrayJavaType arrayJavaType = (ArrayJavaType) type;
+      return isSerializable(arrayJavaType.elementType());
+    }
+    return false;
+  }
+
+  private boolean isParametrized(Type type) {
+    return type instanceof ParametrizedTypeJavaType;
+  }
+
+  private boolean isSerializableParametrized(ParametrizedTypeJavaType type) {
+    // note: this is assuming that custom implementors of Collection
+    // have the good sense to make it serializable just like all implementations in the JDK
+    return (type.isSubtypeOf("java.io.Serializable") || type.isSubtypeOf("java.util.Collection"))
+      && type.typeParameters().stream().allMatch(t -> {
+      JavaType javaType = type.substitution(t);
+      assert javaType != null;
```

Comment:
In production the `assert` statement should have no effect, so users should see the same kind of crash with or without it. It would only be triggered during tests, other than that it's just documentation.

In terms of coverage, for what it's worth, IntelliJ shows the method fully covered when I run `SerializableObjectInSessionCheckTest`.

I removed it anyway because I did not like the tediousness anyway, but I let IntelliJ slap a `//noinspection ConstantConditions` on the statement to make the warning go away, with an explanatory comment.